Added Tag Input Control

mzparacha opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Link to Issue
Closes: https://github.com/hicommonwealth/commonwealth/issues/5085

## Description of Changes
Figma: https://www.figma.com/file/ePuGFatYQI6GL72i27sVwj/Gating?type=design&node-id=431-26125&mode=design&t=x2APs4hUw6AgmqlR-0

- Updated `SelectList` with tag input control styles.
- Added form validation following the [newer forms](https://github.com/hicommonwealth/commonwealth/issues/4150) approach.


## Test Plan
- To see the component, go to `localhost:8080/components`, it will be near the last section of the page (above the forms sections). Try out the usage to see any unexpected behaviour.
- To test the form validation, go to `localhost:8080/components`, and visit the last forms section of the page. There is a form with the tag/select list. Try its usage to see form validation and any unexpected behaviour

## Deployment Plan
N/A

## Other Considerations
This PR is a blcoker to https://github.com/hicommonwealth/commonwealth/pull/5045. The https://github.com/hicommonwealth/commonwealth/pull/5045 PR needs this component as a Form Control

mzparacha commented:
I was initially going to do this as a part of #5045, but I realized that there may be some UX/Logic or other enhancements needed. Also adding this as part of #5045 will clutter the comments section. 

For better management, I decided to do this separately from #5045. But I will merge the changes in this branch to the #5045 to keep things going, and will manage any changes to this PR's code separately (assuming there aren't any breaking changes). It will be all synced when we merge this to master.

cc: @jnaviask @CowMuon

masvelio commented on packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTagInput/CWTagInput.scss:
```diff
@@ -0,0 +1,96 @@
+  .CWTagInput {
+    display: flex;
+    align-items: center;
+    // variable styles (TODO: use css variables)
```
what this relates to?

mzparacha replied:
I forgot to update the CSS to use the SCSS vars we have

masvelio commented on packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTagInput/CWTagInput.scss:
```diff
@@ -0,0 +1,96 @@
+    border-radius: 6px;
+    width: 100%;
+    flex-wrap: wrap;
+    border: 1px solid #E0DFE1;
```
use const instead of E0DFE1

masvelio commented on packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTagInput/CWTagInput.scss:
```diff
@@ -0,0 +1,96 @@
+
+    &:focus-within {
+      // Chrome specific colors
+      border-color: #338FFF;
```
use color from pallet everywhere in this file instead of hex

masvelio commented on packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTagInput/CWTagInput.tsx:
```diff
@@ -0,0 +1,193 @@
+  // fit) - this is a nice to add
+};
+
+const Tag = ({ children, onRemove = () => null, canRemove = true }) => {
```
we already have a tag component that we should reuse => new_designs/cw_tag

masvelio reviewed (commented):
1. I don't like that we are building those types of components from scratch instead of using some libs that cover all those edge cases and accessibility, for example https://mui.com/base-ui/react-autocomplete/ or useAutocomplete. 
2. Second thing is that you did very similar component that @martins87 has done recently with his PR https://github.com/hicommonwealth/commonwealth/pull/4462 . Now we have two alike components with different approach and different implementation. I am not saying that they should be the same component, but we are reinventing the wheel writing again the same logic for passing options, showing tags, handling escape, enter etc. We (eng, design, product) spend DAYS on reviewing Daniel's PR and making sure that everything works fine. 

We should communicate better before building the new components, especially with @Miaplacidus who leads the design system work. Also, we should have identified this friction a way earlier - I mentioned [about the design review](https://github.com/hicommonwealth/commonwealth/issues/4954#issuecomment-1708273420) done by eng before the start of writing gating code.

Not sure about the next steps. Now time is ticking but as I said, we should decide on it way earlier.

masvelio commented:
![image](https://github.com/hicommonwealth/commonwealth/assets/14819225/9fb5f9ed-51d8-4f05-ba25-16a781aec3c8)

1. you can create empty tags

![image](https://github.com/hicommonwealth/commonwealth/assets/14819225/9a627e7d-451b-435e-8b3a-bab34dcbd005)
![image](https://github.com/hicommonwealth/commonwealth/assets/14819225/8e094025-614f-480e-bd6d-d503390b345e)

2. height of empty input is different than height of input with tags

masvelio commented:
we also use https://react-select.com/ in the project already, and this also could be good choice

masvelio commented:
@mzparacha considering that we still have a lot of front end work to do on gating, the design review and tickets review would be necessary for other tickets (possibly breaking up tickets to smaller ones)

mzparacha commented:
Thank you for the feedback @masvelio.
 
I think it all depends on how much

mzparacha commented:
Thank you for the feedback @masvelio 

Just expressing my thoughts here -> I believe the issue here is more of an implementation detail, rather than an expected outcome ex: either using custom comps or third-party comps, we can get the same result from both. I agree that building from scratch is not always ideal, but it does offer more control over the final component. TBH I mostly incline towards custom-tailored solutions for extended control, but that's just me. 

> Second thing is that you did very similar component that @martins87 has done recently with his PR https://github.com/hicommonwealth/commonwealth/pull/4462 .

I agree those are similar, but differ in some ways (as you also mentioned) like the suggestion menu and the search behavior. We can think of this component as the more abstract variant of the search (but depends on how we want this to work based on the above custom/third-party component discussion)

> We should communicate better before building the new components, especially with @Miaplacidus who leads the design system work. Also, we should have identified this friction a way earlier - I mentioned https://github.com/hicommonwealth/commonwealth/issues/4954#issuecomment-1708273420 done by eng before the start of writing gating code.

I agree with this and will try to communicate better next time (😓 apologies to both).

> Not sure about the next steps.

 I only created it this way to focus more on the gating forms stuff (this I will be doing in parallel).

- Since we are deciding on changing the core logic, I will not be adding the requested change for now, until we make a final call there 
- I am happy to move this implementation to use react-select. Please let me know if I should do that @masvelio? and also if you'd like the search component to be based on a common component (we can extract common logic).

mzparacha commented:
Thanks @masvelio. I believe the next step is to use the react-select library, I will be on it in some time.

mzparacha commented:
@masvelio I updated the `SelectList` component with the tag input control styles and removed the custom tag input component. A sample usage would look like this

https://github.com/hicommonwealth/commonwealth/blob/95ee90acd0ec65a669cf862ed741f38ad6e89406/packages/commonwealth/client/scripts/views/components/component_kit/cw_component_showcase.tsx#L198-L210

Would love to hear your thoughts.

masvelio commented on packages/commonwealth/client/scripts/views/components/component_kit/cw_component_showcase.tsx:
```diff
@@ -193,6 +194,21 @@ export const ComponentShowcase = () => {
+          { value: 'binance', label: 'Binance' },
+        ]}
+      />
+      <div style={{ height: '500px' }} />
```
this div is not necessary

mzparacha replied:
Yup, this was just to give space for the dropdown (again its a sample page right?)

mzparacha replied:
Removed

masvelio commented on packages/commonwealth/client/styles/components/component_kit/cw_select_list.scss:
```diff
@@ -2,4 +2,91 @@
+    border: 1px solid $neutral-200 !important;
+    border-radius: 6px !important;
+
+    [class$='-control'] {
```
take a look here https://react-select.com/styles
We already use it in `SelectList` to style some components, so would be good to follow this pattern probably.

mzparacha replied:
Hm, i initially followed that pattern but before committing moved to all CSS, thinking behind that it will be easy to add responsiveness later on.

mzparacha replied:
LMK, what should be the correct approach

mzparacha replied:
> what kind of responsiveness?

screen size responsive styles

> Shouldnt it be always 100% width of the parent container, no matter what device size it is?

I think that depends on the usage

> I am more worried about polluting tsx component with a lot of inline styles 🤔 so maybe we can leave it like that for now

same here

masvelio reviewed (commented):
1. I can see differences in the design vs implementation
![image](https://github.com/hicommonwealth/commonwealth/assets/14819225/624d0a44-091c-4a69-97f4-de58d3c3baab)
![image](https://github.com/hicommonwealth/commonwealth/assets/14819225/18a35dab-db75-4683-bdc2-03ed7141b1bf)

- border color on active state
- input height
- list item background color

Apart from styles, arrow keys does not work (but they do in other SelectList components) 🤔 

2. Other `SelectList` components are broken in styles
![image](https://github.com/hicommonwealth/commonwealth/assets/14819225/eb9e8692-9c39-4a1c-b5e1-273363e52724)
![image](https://github.com/hicommonwealth/commonwealth/assets/14819225/999eeb6e-fb21-451b-b132-51353b6cb730)
![image](https://github.com/hicommonwealth/commonwealth/assets/14819225/7990d808-8e8f-4d7f-89ee-fed462733cd6)

mzparacha commented on packages/commonwealth/client/scripts/views/components/component_kit/cw_component_showcase.tsx:
```diff
@@ -193,6 +194,21 @@ export const ComponentShowcase = () => {
+          { value: 'substrate', label: 'Substrate' },
+          { value: 'binance', label: 'Binance' },
+        ]}
+      />
```
Done

mzparacha commented:
About the design, it's using some base styles from the existing `SelectList`, that component is used in a lot of existing areas.

I think the correct approach would be to add a new component `SelectList` to the `new_desgins` folder and add the new styles there as we have done with a lot of existing components and maybe we can migrate existing `SelectList` usage to the new styles at a later time (since it's not the scope of this ticket).

Lmk your thoughts @masvelio

masvelio commented:
> About the design, it's using some base styles from the existing `SelectList`, that component is used in a lot of existing areas.
> 
> I think the correct approach would be to add a new component `SelectList` to the `new_desgins` folder and add the new styles there as we have done with a lot of existing components and maybe we can migrate existing `SelectList` usage to the new styles at a later time (since it's not the scope of this ticket).
> 
> Lmk your thoughts @masvelio

I am ok with working on styles on separate ticket. But make sure that the styles are not broken in other places as I mentioned and showed above in the screenshots.

mzparacha commented:
Ready for another round of review. @masvelio.

I also added the newer form validation for the select list in this PR, and a sample form in the showcase page (`localhost:3000/components`) containing the select list.

masvelio commented:
![image](https://github.com/hicommonwealth/commonwealth/assets/14819225/ac4d85ac-5fc6-421a-abda-4d493aee14b9)

I have external mouse plugged in and I can see vertical and horizontal scrollbar, even though it is not scrollable.

masvelio commented on packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWSelectList/CWSelectList.scss:
```diff
@@ -0,0 +1,108 @@
+        margin: 0px !important;
+        margin-right: 8px !important;
+        background-color: $neutral-100;
+        // max-height: 24px;
```
is this necessary? if not, please remove

mzparacha replied:
nice spot, will remove shortly

mzparacha replied:
[Done](https://github.com/hicommonwealth/commonwealth/pull/5086/commits/1ad60ace7ac529b5378418402efe2995001f0f2b)

masvelio commented on packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWSelectList/CWSelectList.tsx:
```diff
@@ -0,0 +1,123 @@
+  }, [props.hookToForm, props.name, props.value, formContext]);
+
+  return (
+    <div>
```
maybe `CWSelectList` class would be good to add here, will be easier to find this component in the DOM

mzparacha replied:
i agree, will add shortly

mzparacha replied:
[Done](https://github.com/hicommonwealth/commonwealth/pull/5086/commits/718240a20d980f0515f69f15e0eeee8aa2e7f459)

masvelio reviewed (approved):
1. So now we have `SelectList` and `CWSelectList`. Could you explain the difference between them? Maybe we need another ticket to consolidate them? I am assuming that we should use `CWSelectList` everywhere and get rid of `SelectList`?

2. Using arrow does not work in your component but it works fine in `SelectList` eg in topic selection. Is this expected?
![image](https://github.com/hicommonwealth/commonwealth/assets/14819225/a720c2d9-3c05-4bd1-b404-d69447f2d1ac)

Other than that, it generally looks and feels great! I would merge it to unblock your work without design sign-off for now. They can do the review by the end of gating in a more holistic way. This is my opinion, I would leave the decision to @jnaviask

Miaplacidus commented:
@mzparacha 
My biggest concern was the similarity between this component and @martins87 search bar component (of course, they have different purposes). Having said that, I think the differences between them justifies two separate components, i.e. consolidation is not the best choice. However, since we've opted to use react-select for this card, it would be internally consistent for us to use it for the search bar component if possible (thoughts, @martins87?), but that's a matter external to this card.  

In regards to SelectList vs CWSelectList, consolidation of these is also going to have to be dealt with outside of this card.

jnaviask commented:
cc @mzparacha to address Marcin's comments, then will review / merge.

mzparacha commented:
> 1. So now we have SelectList and CWSelectList. Could you explain the difference between them? Maybe we need another ticket to consolidate them? I am assuming that we should use CWSelectList everywhere and get rid of SelectList?
> @mzparacha
My biggest concern was the similarity between this component and @martins87 search bar component (of course, they have different purposes). Having said that, I think the differences between them justifies two separate components, i.e. consolidation is not the best choice. However, since we've opted to use react-select for this card, it would be internally consistent for us to use it for the search bar component if possible (thoughts, @martins87?), but that's a matter external to this card.
In regards to SelectList vs CWSelectList, consolidation of these is also going to have to be dealt with outside of this card.

I agree with both comments. I also saw that we have a related ticket https://github.com/hicommonwealth/commonwealth/issues/5142, I think we can do consolidation in that ticket. Thoughts?

> 2. Using arrow does not work in your component but it works fine in SelectList eg in topic selection. Is this expected?

I believe its due to the multi value (we allow multi value for select). But still an area to investigate. I will leave an update shortly.

Other than that, its ready for merge.

mzparacha commented:
> I believe its due to the multi value (we allow multi value for select). But still an area to investigate. I will leave an update shortly.

Turns out it was just a color issue, the focus was working but not visible. Fixed [here](https://github.com/hicommonwealth/commonwealth/pull/5086/commits/f249317f35b91280e043b67a9e83d1d024712e7c).

Ready for final review.

cc: @jnaviask @masvelio

masvelio commented:
🟢

jnaviask reviewed (changes requested):
When I enter an invalid string and hit "enter" twice in the form, the error message appears then disappears. The error message should persist until the error is corrected.

mzparacha commented:
@jnaviask Not sure what you mean, is it something only happening with the last form field? sidenote: pressing enter on the last field submits the form

mzparacha commented:
> When I enter an invalid string and hit "enter" twice in the form, the error message appears then disappears. The error message should persist until the error is corrected.

@jnaviask I think the issue you faced was with this component (tag-input/select-list) when used in the gating form? If yes, then I think we can fix it in the gating PR and merge this to unblock @Miaplacidus's work on https://github.com/hicommonwealth/commonwealth/pull/5169?

jnaviask reviewed (approved):
Let's get this into v0.7.0, as it's a blocker.
